routing/missioncontrol_store.go

```diff
@@ -0,0 +1,235 @@
+type missionControlStore struct {
+	db         *bbolt.DB
+	maxRecords int
+}
+
+func newMissionControlStore(db *bbolt.DB, maxRecords int) (*missionControlStore, error) {
+	// Create buckets if not yet existing.
+	err := db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(resultsKey)
+		if err != nil {
+			return fmt.Errorf("cannot create results bucket: %v",
+				err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &missionControlStore{
+		db:         db,
+		maxRecords: maxRecords,
+	}, nil
+}
+
+// clear removes all results from the db.
+func (b *missionControlStore) clear() error {
+	return b.db.Update(func(tx *bbolt.Tx) error {
+		if err := tx.DeleteBucket(resultsKey); err != nil {
+			return err
+		}
+
+		_, err := tx.CreateBucket(resultsKey)
+		return err
+	})
+}
+
+// fetchAll returns all results currently stored in the database.
+func (b *missionControlStore) fetchAll() ([]*paymentResult, error) {
+	var results []*paymentResult
+
+	err := b.db.View(func(tx *bbolt.Tx) error {
+		resultBucket := tx.Bucket(resultsKey)
+		results = make([]*paymentResult, resultBucket.Stats().KeyN)
+
+		i := 0
+		return resultBucket.ForEach(func(k, v []byte) error {
+			result, err := deserializeResult(k, v)
+			if err != nil {
+				return err
+			}
+
+			results[i] = result
+			i++
+
+			return nil
+		})
+
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return results, nil
+}
+
+// deserializeResult deserializes a payment result.
+func deserializeResult(k, v []byte) (
+	*paymentResult, error) {
+
+	// Parse payment id.
+	result := paymentResult{
+		id: byteOrder.Uint64(k[8:]),
+	}
+
+	r := bytes.NewReader(v)
+
+	// Read timestamps, success status and failure source index.
+	var (
+		timeFwd, timeReply uint64
+		dbFailureSourceIdx int32
+	)
+
+	err := channeldb.ReadElements(
+		r, &timeFwd, &timeReply, &result.success, &dbFailureSourceIdx,
+	)
+	if err != nil {
+		return nil, err
+	}
+	result.timeFwd = time.Unix(0, int64(timeFwd)).UTC()
+	result.timeReply = time.Unix(0, int64(timeReply)).UTC()
+	if dbFailureSourceIdx != -1 {
+		failureSourceIdx := int(dbFailureSourceIdx)
+		result.failureSourceIdx = &failureSourceIdx
+	}
+
+	// Read route.
+	route, err := channeldb.DeserializeRoute(r)
+	if err != nil {
+		return nil, err
+	}
+	result.route = &route
+
+	// Read failure.
+	failureBytes, err := wire.ReadVarBytes(
+		r, 0, lnwire.FailureMessageLength, "failure",
+	)
+	if err != nil {
+		return nil, err
+	}
+	if len(failureBytes) > 0 {
+		result.failure, err = lnwire.DecodeFailureMessage(
+			bytes.NewReader(failureBytes), 0,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &result, nil
+}
+
+// AddResult adds a new result to the db.
+func (b *missionControlStore) AddResult(rp *paymentResult) error {
+	return b.db.Update(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(resultsKey)
+
+		// Prune oldest entries.
+		if b.maxRecords > 0 {
+			count := bucket.Stats().KeyN
```

Comment:
we've had issues in the past using `Stats()`, perhaps instead we can count the number of entries in the db on startup and track the current count in memory?

Reply:
Do you have more details on what issues those were? We can work around it, but it would be nice to have some confirmation that `Stats().KeyN` is indeed unreliable.

Reply:
it caused a panic in one of our prior migrations:
 * issue: https://github.com/lightningnetwork/lnd/issues/1874#issuecomment-420022023
 * fix: https://github.com/lightningnetwork/lnd/pull/1898

the root cause was never determined, so we've been avoiding `Stats()` like the plague

Reply:
Ah, I see. So there is no other way to get a count than iterating through all items with a cursor?

Reply:
Ah, I see. So there is no other way to get a count than iterating through all items with a cursor?

Reply:
The issue that you mention, could it be that it was caused by a database that was already corrupt?

Reply:
> Ah, I see. So there is no other way to get a count than iterating through all items with a cursor?

yeah that's the simplest way

> The issue that you mention, could it be that it was caused by a database that was already corrupt?

we saw it happen on perfectly good dbs. once the fix landed, people were able to resume w/o issue afaiaa

Reply:
`Stats()` call removed. Small complication because store needs to be idempotent, but solved that with an additional bucket get.

How did you actually know these were perfectly good dbs?